<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2025-10-15" />

<title>Introduction to Data Visualization</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">data_visualization_in_R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Introduction to Data Visualization</h1>
<h4 class="date">2025-10-15</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-10-15
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>data_visualization_in_R/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown is untracked by Git. To know which version of the R
Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20251007code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20251007)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20251007code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20251007)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomcrazyhottommydatavisualizationinRtree9e1856ebf86b04f5c4ee85a9ad613046b1037c42targetblank9e1856ea">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/crazyhottommy/data_visualization_in_R/tree/9e1856ebf86b04f5c4ee85a9ad613046b1037c42" target="_blank">9e1856e</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomcrazyhottommydatavisualizationinRtree9e1856ebf86b04f5c4ee85a9ad613046b1037c42targetblank9e1856ea"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/crazyhottommy/data_visualization_in_R/tree/9e1856ebf86b04f5c4ee85a9ad613046b1037c42" target="_blank">9e1856e</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  analysis/01_intro_data_viz.Rmd
    Untracked:  analysis/02_intro_to_ggplot2.Rmd
    Untracked:  analysis/03_heatmap_demystified.Rmd
    Untracked:  analysis/04_practical_scRNAseq_viz.Rmd

Unstaged changes:
    Modified:   analysis/about.Rmd
    Modified:   analysis/index.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<div id="the-power-of-data-visualization" class="section level1">
<h1>The Power of Data Visualization</h1>
<p>Data visualization is the graphical representation of information and
data. In an age where we generate data at an unprecedented rate, the
ability to create clear, compelling visualizations has become crucial
for understanding patterns, communicating findings, and making
data-driven decisions.</p>
<blockquote>
<p>“A picture is worth a thousand words” - especially when that picture
represents complex data patterns that would be nearly impossible to
detect in tables of numbers.</p>
</blockquote>
<div id="why-visualize-data" class="section level2">
<h2>Why Visualize Data?</h2>
<div id="pattern-recognition" class="section level3">
<h3>1. <strong>Pattern Recognition</strong></h3>
<p>Human brains are exceptionally good at processing visual information.
We can quickly spot trends, outliers, and relationships in graphical
data that might take hours to detect in numerical tables.</p>
</div>
<div id="communication" class="section level3">
<h3>2. <strong>Communication</strong></h3>
<p>Visualizations can communicate complex findings to diverse audiences
- from fellow researchers to policymakers to the general public.</p>
</div>
<div id="exploration" class="section level3">
<h3>3. <strong>Exploration</strong></h3>
<p>Visualization is a powerful tool for exploratory data analysis,
helping researchers discover unexpected patterns and generate new
hypotheses.</p>
</div>
<div id="validation" class="section level3">
<h3>4. <strong>Validation</strong></h3>
<p>Visual inspection can help identify data quality issues, outliers,
and assumptions that might not be obvious from summary statistics
alone.</p>
</div>
</div>
<div id="the-genomics-visualization-landscape" class="section level2">
<h2>The Genomics Visualization Landscape</h2>
<p>If you read genomics papers regularly, you’ll notice that most
figures fall into just six categories:</p>
<ol style="list-style-type: decimal">
<li><strong>Scatter plots</strong> - for correlation analysis and gene
expression relationships</li>
<li><strong>Bar plots</strong> - for comparing quantities across
groups</li>
<li><strong>Line plots</strong> - for showing trends over time or
conditions</li>
<li><strong>Box plots/Violin plots</strong> - for comparing
distributions across groups</li>
<li><strong>Histograms</strong> - for understanding data
distributions</li>
<li><strong>Heatmaps</strong> - for visualizing large matrices of data
(gene expression, etc.)</li>
</ol>
<p>Mastering these six types will enable you to reproduce approximately
90% of figures in genomics literature.</p>
</div>
<div id="principles-of-effective-data-visualization"
class="section level2">
<h2>Principles of Effective Data Visualization</h2>
<div id="know-your-audience" class="section level3">
<h3>1. <strong>Know Your Audience</strong></h3>
<ul>
<li><strong>Scientific peers</strong>: Detailed, precise, with
statistical information</li>
<li><strong>General public</strong>: Simple, clear, with minimal
jargon</li>
<li><strong>Policymakers</strong>: Focused on actionable insights</li>
</ul>
</div>
<div id="choose-the-right-chart-type" class="section level3">
<h3>2. <strong>Choose the Right Chart Type</strong></h3>
<table>
<colgroup>
<col width="26%" />
<col width="33%" />
<col width="40%" />
</colgroup>
<thead>
<tr class="header">
<th>Data Type</th>
<th>Relationship</th>
<th>Best Chart Type</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Continuous vs Continuous</td>
<td>Correlation</td>
<td>Scatter plot</td>
</tr>
<tr class="even">
<td>Categorical vs Continuous</td>
<td>Distribution comparison</td>
<td>Box plot, Violin plot</td>
</tr>
<tr class="odd">
<td>Single continuous variable</td>
<td>Distribution</td>
<td>Histogram</td>
</tr>
<tr class="even">
<td>Categories vs Quantities</td>
<td>Comparison</td>
<td>Bar plot</td>
</tr>
<tr class="odd">
<td>Time series</td>
<td>Trends</td>
<td>Line plot</td>
</tr>
<tr class="even">
<td>Matrix data</td>
<td>Patterns in 2D</td>
<td>Heatmap</td>
</tr>
</tbody>
</table>
</div>
<div id="design-principles" class="section level3">
<h3>3. <strong>Design Principles</strong></h3>
<div id="clarity-over-beauty" class="section level4">
<h4><strong>Clarity Over Beauty</strong></h4>
<ul>
<li>Prioritize understanding over aesthetics</li>
<li>Remove unnecessary elements (chartjunk)</li>
<li>Use consistent styling</li>
</ul>
</div>
<div id="accessibility" class="section level4">
<h4><strong>Accessibility</strong></h4>
<ul>
<li>Use colorblind-friendly palettes</li>
<li>Include proper legends and labels</li>
<li>Ensure sufficient contrast</li>
</ul>
</div>
<div id="accuracy" class="section level4">
<h4><strong>Accuracy</strong></h4>
<ul>
<li>Start axes at zero when appropriate</li>
<li>Use appropriate scales</li>
<li>Don’t distort data relationships</li>
</ul>
</div>
</div>
<div id="color-usage" class="section level3">
<h3>4. <strong>Color Usage</strong></h3>
<div id="sequential-colors-for-ordered-data" class="section level4">
<h4><strong>Sequential Colors</strong> (for ordered data)</h4>
<ul>
<li>Light to dark of single hue</li>
<li>Good for: heatmaps, choropleth maps</li>
<li>Example: gene expression levels</li>
</ul>
</div>
<div id="diverging-colors-for-data-with-meaningful-center"
class="section level4">
<h4><strong>Diverging Colors</strong> (for data with meaningful
center)</h4>
<ul>
<li>Two hues meeting at neutral center</li>
<li>Good for: correlation matrices, fold changes</li>
<li>Example: up/down regulated genes</li>
</ul>
</div>
<div id="categorical-colors-for-distinct-groups" class="section level4">
<h4><strong>Categorical Colors</strong> (for distinct groups)</h4>
<ul>
<li>Distinct, equally prominent hues</li>
<li>Good for: different cancer types, treatment groups</li>
<li>Avoid: red/green combinations (colorblind unfriendly)</li>
</ul>
</div>
</div>
</div>
<div id="common-visualization-mistakes" class="section level2">
<h2>Common Visualization Mistakes</h2>
<div id="the-misleading-y-axis" class="section level3">
<h3>1. <strong>The Misleading Y-Axis</strong></h3>
<p>Starting bar charts at non-zero values can exaggerate
differences.</p>
</div>
<div id="too-many-categories" class="section level3">
<h3>2. <strong>Too Many Categories</strong></h3>
<p>More than 7-8 categories become difficult to distinguish, especially
with color.</p>
</div>
<div id="pie-chart-overuse" class="section level3">
<h3>3. <strong>Pie Chart Overuse</strong></h3>
<p>Pie charts are difficult to interpret precisely. Bar charts are
usually better.</p>
</div>
<div id="d-effects" class="section level3">
<h3>4. <strong>3D Effects</strong></h3>
<p>3D charts are often harder to read and can distort data
perception.</p>
</div>
<div id="poor-color-choices" class="section level3">
<h3>5. <strong>Poor Color Choices</strong></h3>
<ul>
<li>Using red/green for comparisons (colorblind issues)</li>
<li>Not enough contrast</li>
<li>Too many similar colors</li>
</ul>
</div>
</div>
<div id="the-grammar-of-graphics" class="section level2">
<h2>The Grammar of Graphics</h2>
<p>The <strong>Grammar of Graphics</strong> is a theoretical framework
for data visualization that underlies ggplot2. It breaks down
visualizations into fundamental components:</p>
<div id="core-components" class="section level3">
<h3>Core Components:</h3>
<ol style="list-style-type: decimal">
<li><strong>Data</strong>: The dataset you want to visualize</li>
<li><strong>Aesthetics (aes)</strong>: How data variables map to visual
properties
<ul>
<li>Position (x, y)</li>
<li>Color, fill</li>
<li>Size, shape</li>
<li>Transparency (alpha)</li>
</ul></li>
<li><strong>Geometries (geom)</strong>: The visual elements that
represent data
<ul>
<li>Points, lines, bars, boxes</li>
</ul></li>
<li><strong>Statistics (stat)</strong>: Statistical transformations of
data
<ul>
<li>Counts, means, regression lines</li>
</ul></li>
<li><strong>Scales</strong>: Control how aesthetics map to data values
<ul>
<li>Color scales, axis scales</li>
</ul></li>
<li><strong>Coordinate systems</strong>: How data is mapped to the plane
<ul>
<li>Cartesian, polar, map projections</li>
</ul></li>
<li><strong>Facets</strong>: Subplots for different subsets of data</li>
<li><strong>Themes</strong>: Overall visual appearance</li>
</ol>
</div>
<div id="the-layered-approach" class="section level3">
<h3>The Layered Approach</h3>
<p>ggplot2 builds visualizations by adding layers:</p>
<pre class="r"><code>ggplot(data = mydata, aes(x = var1, y = var2)) +  # Base layer
  geom_point() +                                   # Geometry layer
  geom_smooth(method = &quot;lm&quot;) +                     # Additional geometry
  scale_color_brewer(type = &quot;qual&quot;) +              # Scale layer
  theme_minimal() +                                # Theme layer
  labs(title = &quot;My Plot&quot;)                          # Labels layer</code></pre>
<p>This modular approach makes ggplot2 extremely flexible and
powerful.</p>
</div>
</div>
<div id="data-visualization-in-scientific-research"
class="section level2">
<h2>Data Visualization in Scientific Research</h2>
<div id="pre-analysis-visualization" class="section level3">
<h3>Pre-Analysis Visualization</h3>
<ul>
<li><strong>Data exploration</strong>: Understanding distributions,
identifying outliers</li>
<li><strong>Quality control</strong>: Spotting data collection
issues</li>
<li><strong>Hypothesis generation</strong>: Discovering unexpected
patterns</li>
</ul>
</div>
<div id="analysis-visualization" class="section level3">
<h3>Analysis Visualization</h3>
<ul>
<li><strong>Results presentation</strong>: Showing statistical
findings</li>
<li><strong>Model diagnostics</strong>: Checking assumptions,
residuals</li>
<li><strong>Sensitivity analysis</strong>: Exploring robustness of
findings</li>
</ul>
</div>
<div id="communication-visualization" class="section level3">
<h3>Communication Visualization</h3>
<ul>
<li><strong>Publication figures</strong>: Journal-ready
visualizations</li>
<li><strong>Presentations</strong>: Conference talks, seminars</li>
<li><strong>Reports</strong>: Technical documentation</li>
</ul>
</div>
</div>
<div id="best-practices-for-scientific-visualization"
class="section level2">
<h2>Best Practices for Scientific Visualization</h2>
<div id="reproducibility" class="section level3">
<h3>1. <strong>Reproducibility</strong></h3>
<ul>
<li>Use code-based tools (like R) rather than point-and-click
software</li>
<li>Document your visualization code</li>
<li>Use version control for your analysis scripts</li>
</ul>
</div>
<div id="statistical-honesty" class="section level3">
<h3>2. <strong>Statistical Honesty</strong></h3>
<ul>
<li>Show uncertainty (error bars, confidence intervals)</li>
<li>Don’t cherry-pick flattering visualizations</li>
<li>Include sample sizes and statistical tests when relevant</li>
</ul>
</div>
<div id="ethical-considerations" class="section level3">
<h3>3. <strong>Ethical Considerations</strong></h3>
<ul>
<li>Don’t manipulate scales to exaggerate effects</li>
<li>Be transparent about data filtering or transformations</li>
<li>Consider the broader implications of your visualizations</li>
</ul>
</div>
</div>
<div id="tools-and-resources" class="section level2">
<h2>Tools and Resources</h2>
<div id="r-ecosystem-for-visualization" class="section level3">
<h3>R Ecosystem for Visualization</h3>
<ul>
<li><strong>ggplot2</strong>: The gold standard for statistical
graphics</li>
<li><strong>plotly</strong>: Interactive visualizations</li>
<li><strong>gganimate</strong>: Animated plots</li>
<li><strong>patchwork</strong>: Combining multiple plots</li>
</ul>
</div>
<div id="color-resources" class="section level3">
<h3>Color Resources</h3>
<ul>
<li><strong>ColorBrewer</strong>: Scientifically-designed color
palettes</li>
<li><strong>Viridis</strong>: Perceptually uniform, colorblind-friendly
palettes</li>
<li><strong>RColorBrewer</strong>: R implementation of ColorBrewer</li>
</ul>
</div>
<div id="inspiration-and-learning" class="section level3">
<h3>Inspiration and Learning</h3>
<ul>
<li><a href="https://www.r-graph-gallery.com/">R Graph Gallery</a>:
Hundreds of chart examples</li>
<li><a href="https://www.data-to-viz.com/">From Data to Viz</a>: Choose
the right chart type</li>
<li><a href="http://www.storytellingwithdata.com/">Storytelling with
Data</a>: Effective communication principles</li>
</ul>
</div>
</div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<p>Essential reading for mastering data visualization:</p>
<ol style="list-style-type: decimal">
<li><strong><a href="https://clauswilke.com/dataviz/">Fundamentals of
Data Visualization</a></strong> by Claus O. Wilke
<ul>
<li>Comprehensive guide to visualization principles</li>
<li>Covers theory and practical applications</li>
<li>Excellent for understanding when to use different chart types</li>
</ul></li>
<li><strong><a href="https://socviz.co/">Data Visualization</a></strong>
by Kieran Healy
<ul>
<li>Practical approach with R and ggplot2</li>
<li>Focus on social science applications</li>
<li>Great for learning ggplot2 syntax and concepts</li>
</ul></li>
<li><strong><a href="https://ggplot2-book.org/">ggplot2: Elegant
Graphics for Data Analysis</a></strong> by Hadley Wickham
<ul>
<li>The definitive guide to ggplot2</li>
<li>Written by the package creator</li>
<li>Essential reference for advanced ggplot2 techniques</li>
</ul></li>
</ol>
<div id="next-steps" class="section level2">
<h2>Next Steps</h2>
<p>Now that you understand the principles of effective data
visualization, you’re ready to start creating your own visualizations.
In the next lesson, we’ll dive into practical ggplot2 programming using
real cancer genomics data.</p>
<p><strong>Continue to:</strong> <a
href="02_intro_to_ggplot2.html">Lesson 2: Practical ggplot2</a></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.1 (2024-06-14)
Platform: aarch64-apple-darwin20
Running under: macOS Sonoma 14.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] vctrs_0.6.5       httr_1.4.7        cli_3.6.3         knitr_1.48       
 [5] rlang_1.1.4       xfun_0.52         stringi_1.8.4     processx_3.8.4   
 [9] promises_1.3.0    jsonlite_1.8.8    glue_1.8.0        rprojroot_2.0.4  
[13] git2r_0.35.0      htmltools_0.5.8.1 httpuv_1.6.15     ps_1.7.7         
[17] sass_0.4.9        fansi_1.0.6       rmarkdown_2.27    jquerylib_0.1.4  
[21] tibble_3.2.1      evaluate_0.24.0   fastmap_1.2.0     yaml_2.3.10      
[25] lifecycle_1.0.4   whisker_0.4.1     stringr_1.5.1     compiler_4.4.1   
[29] fs_1.6.4          pkgconfig_2.0.3   Rcpp_1.0.13       rstudioapi_0.16.0
[33] later_1.3.2       digest_0.6.36     R6_2.5.1          utf8_1.2.4       
[37] pillar_1.9.0      callr_3.7.6       magrittr_2.0.3    bslib_0.8.0      
[41] tools_4.4.1       cachem_1.1.0      getPass_0.2-4    </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
